internal/messagequeue/messagequeue.go

```diff
@@ -588,41 +590,27 @@ func (mq *MessageQueue) extractOutgoingMessage(supportsHave bool) (bsmsg.BitSwap
 			wantType = pb.Message_Wantlist_Block
 		}
 
-		e := bcstEntries[i]
+		e := bcstEntries[bcstSentCount]
 		msgSize += mq.msg.AddEntry(e.Cid, e.Priority, wantType, false)
 	}
 
 	// Called when the message has been successfully sent.
-	onMessageSent := func(wantlist []bsmsg.Entry) {
-		bcst := keysToSet(bcstEntries)
-		prws := keysToSet(peerEntries)
-
+	onMessageSent := func() {
 		mq.wllock.Lock()
 		defer mq.wllock.Unlock()
 
 		// Move the keys from pending to sent
-		for _, e := range wantlist {
-			if _, ok := bcst[e.Cid]; ok {
-				mq.bcstWants.Sent(e)
-			}
-			if _, ok := prws[e.Cid]; ok {
-				mq.peerWants.Sent(e)
-			}
+		for i := 0; i < bcstSentCount; i++ {
+			mq.bcstWants.MarkSent(bcstEntries[i])
+		}
+		for i := 0; i < peerSentCount; i++ {
+			mq.peerWants.MarkSent(peerEntries[i])
```

Comment:
This won't quite work as we will have removed some of the peer wants to peers that don't support HAVE.

Reply:
We remove them when we realize the peer doesn't support HAVE, so in `MarkSent()` we could just check if they've already been removed from pending before adding them to sent.
But actually I think it's clearer to put them in a set. I pushed a commit with this change.